### PR TITLE
Added tests for JsonLD class

### PR DESCRIPTION
**Pull request by annando**
These are my first tests - more to come (the function `fetchElementArray` needs some tests)

Let's see if the tests are called at all.

### CODE REVIEW

File: tests/src/Util/JSonLDTest.php [hunk 1]
```diff
@@ -0,0 +1,80 @@
...
+		$data = JsonLD::fetchElement($object, 'field');
+		$this->assertSame('value', $data);
+	}
+
+	public function testFetchElementFoundID()
+	{
+		$object = ['field' => ['field2' => 'value2', '@id' => 'value', 'field3' => 'value3']];
+
+		$data = JsonLD::fetchElement($object, 'field');
+		$this->assertSame('value', $data);
```
[the remark this answers is not in the document]
Comment (annando) [reply tests/src/Util/JSonLDTest.php:61]: No. When you - for example - want to fetch the actor then the actor can be a simple string value - or a complete array structure where the field that you want to fetch is the one with the third parameter of `fetchElement` as index. The value *could* be in that index structure - or it *could* be just a string. Both is valid in JsonLD and both is used.

You never know what implementers are doing, so you need to handle both situations with the same function call.

**Comment by annando** [comment]
Hmm, why isn't there any test popping up after the last push?

File: tests/src/Util/JSonLDTest.php (hunk 1 above)
[the remark this answers is not in the document]
Comment (MrPetovan) [reply tests/src/Util/JSonLDTest.php:70]: What if there's no `@id` property?

**Comment by annando** [comment]
The tests are here: https://travis-ci.org/github/friendica/friendica/builds/695419775

File: tests/src/Util/JSonLDTest.php (hunk 1 above)
[the remark this answers is not in the document]
Comment (annando) [reply tests/src/Util/JSonLDTest.php:97]: You don't want to have the whole element. You could - if you wanted to - not just add the string with the receivers in the `to` array, but you could add complete arrays with the whole data that you know about this person (like name, avatar and the whole stuff). Technically this is possible. At some time Hubzilla did so as well. But it doesn't make no sense parsing all these fields.

Same goes with shared posts. All we use from a shared post is the ID of that post. We then fetch the content by ourself. We don't trust the provided data for security reasons.

[the remark this answers is not in the document]
Comment (MrPetovan) [reply tests/src/Util/JSonLDTest.php:97]: Does it mean with `type: Page` attachments we have to fetch the target page ourselves and disregard the title and image provided?

[the remark this answers is not in the document]
Comment (annando) [reply tests/src/Util/JSonLDTest.php:97]: Well, the attachments aren't fetched by the function `fetchElement`. The attachments are fetched by `fetchElementArray` and then are processed element by element.

But yes, we possibly have to think about not trusting provided data that much. But this is an unrelated discussion.

[the remark this answers is not in the document]
Comment (MrPetovan) [reply tests/src/Util/JSonLDTest.php:97]: I'm asking because `fetchElementArray` had the same behavior (returning the `@id` value only by default) as `fetchElement`.

[the remark this answers is not in the document]
Comment (annando) [reply tests/src/Util/JSonLDTest.php:97]: See the test function `testFetchElementArrayFoundArrays()`. The function will return the array.

File: tests/src/Util/JSonLDTest.php [hunk 2]
```diff
@@ -0,0 +1,116 @@
...
+
+		$data = JsonLD::fetchElementArray($object, 'field');
+		$this->assertSame([], $data);
+	}
+
+	public function testFetchElementArrayFoundID()
+	{
+		$object = ['field' => ['value1', ['@id' => 'value2'], ['@id' => 'value3']]];
+
+		$data = JsonLD::fetchElementArray($object, 'field', '@id');
```
Comment (MrPetovan) [inline tests/src/Util/JSonLDTest.php:52]: This is the same call as line 64, it is confusing why you would explicitly mention the default value of the `$key` parameter.

File: tests/src/Util/JSonLDTest.php [hunk 3]
```diff
@@ -0,0 +1,116 @@
...
+		$this->assertSame(['value1', 'value2', 'value3'], $data);
+	}
+
+	public function testFetchElementArrayFoundArrays()
+	{
+		$object = ['field' => [['subfield11' => 'value11', 'subfield12' => 'value12'],
```
Comment (MrPetovan) [inline tests/src/Util/JSonLDTest.php:58]: What if one of these have an `@id`? The output become a array with a single element containing the value of the `@id` property. Is it intended? Why are the other elements discarded **by default**?

Comment (annando) [reply tests/src/Util/JSonLDTest.php:58]: This can't happen with the data, but I just changed the functionality a little bit.

File: tests/src/Util/JSonLDTest.php (hunk 2 above)
Comment (annando) [reply tests/src/Util/JSonLDTest.php:52]: I slightly changed the functionality now.

File: tests/src/Util/JSonLDTest.php (hunk 3 above)
Comment (MrPetovan) [reply tests/src/Util/JSonLDTest.php:58]: What data? The data you made up here? I'm talking about it because I encountered the case working with Mammoth messages. its was compacted alright and the result didn't have `@id` attributes on all the elements I was retrieving.

File: src/Util/JsonLD.php [hunk 4]
```diff
@@ -191,12 +191,10 @@ public static function fetchElementArray($array, $element, $key = '@id')
 		$elements = [];
 
 		foreach ($array[$element] as $entry) {
-			if (!is_array($entry)) {
+			if (!is_array($entry) || (is_null($key) && is_array($entry))) {
 				$elements[] = $entry;
-			} elseif (isset($entry[$key])) {
+			} elseif (!is_null($key) && isset($entry[$key])) {
```
Comment (MrPetovan) [inline src/Util/JsonLD.php:196]: Please swap the conditions if `$key` value must be checked first. Also why do you have to check that `$entry` isn't an array or it is discarded?

File: tests/src/Util/JSonLDTest.php (hunk 3 above)
Comment (annando) [reply tests/src/Util/JSonLDTest.php:58]: Well, "Mammoth" seems to make stuff slightly different than other implementations. Hopefully with my small changes this works better.

File: src/Util/JsonLD.php (hunk 4 above)
Comment (annando) [reply src/Util/JsonLD.php:196]: When no `$key` is present, then all content is taken. When `$key` is present, then only entries with that key are taken - or stuff that isn't an array.

See the test `testFetchElementArrayFoundID`. Depending on the implementers the data could be "flat" or inside some array.

**Comment by annando** [comment]
This PR is blocking https://github.com/friendica/friendica/pull/8727

I guess I now added tests for most (if not all) possible cases.

**Comment by annando** [comment]
These functions are solely for fetching data *directly* from the JsonLD content. So they are for situations where the exact structure isn't known. So it helps with not having the need to check if a specific value exists prior to assign them. Or it helps with: "Give me the content of the content field in the source array when the content-type is 'text/bbcode'". So some kind of very simple XPath.

When you know that the array value exists, then just use the regular array functions. I think the best is to just look at the source.

**Comment by MrPetovan** [comment]
Ok, I've found the case where I need `fetchElementArray` because a list of a single object is compacted into just the object itself, which fools direct array access expecting for a list. And it works as expected with your changes.

**Comment by annando** [comment]
Yeah, thanks to your hints I tested the function and found out that there had been weird edge cases.

**Comment by MrPetovan** [comment]
TESTS FTW!

**Comment by annando** [comment]
Yeah, I hope @nupplaphil is proud of me :-)

**Comment by MrPetovan** [comment]
Wherever he is right now, I'm sure he is proud of you.
